### PR TITLE
Support unparsing Array plan to SQL string

### LINKED ISSUE #11144: Support unparsing the Value Plan of Array (List) to SQL String
### Is your feature request related to a problem or challenge?

In `unparser/expr.rs`, list scalar values haven't been supported yet.
```rust
ScalarValue::FixedSizeList(_a) => not_impl_err!("Unsupported scalar: {v:?}"),
ScalarValue::List(_a) => not_impl_err!("Unsupported scalar: {v:?}"),
ScalarValue::LargeList(_a) => not_impl_err!("Unsupported scalar: {v:?}"),
```
However, I found some problems with supporting it. Currently, array construction is implemented by a `ScalarUDF` called `make_array`.
https://github.com/apache/datafusion/blob/ff116c3da69897358f210a3ea944c8e51dcb7b52/datafusion/sql/src/expr/value.rs#L145-L151
This might mean `ScalarValue::List` is never created from an AST expression node.

If I try to create the logical plan without the default SessionContext, like this:
```rust
let sql = r#"select [1,2,3]"#;

println!("SQL: {:?}", sql);
println!("********");
// parse the SQL
let dialect = GenericDialect {}; // or AnsiDialect, or your own dialect ...
let ast = Parser::parse_sql(&dialect, sql).unwrap();
let statement = &ast[0];
println!("AST: {:?}", statement);

// create a logical query plan
let context_provider = MyContextProvider::new();
let sql_to_rel = SqlToRel::new(&context_provider);
let plan = match sql_to_rel.sql_statement_to_plan(statement.clone()) {
    Ok(plan) => plan,
    Err(e) => {
        println!("Error: {:?}", e);
        return;
    }
};
// show the planned SQL
let planned = match plan_to_sql(&plan) {
    Ok(sql) => sql,
    Err(e) => {
        println!("Error: {:?}", e);
        return;
    }
};

println!("unparsed to SQL:\n {}", planned);
```
I will get the error message:
```
Error: NotImplemented("array_expression feature is disabled, so you should implement the make_array UDF yourself")
```
Then, I tried to add `make_array_udf` in `MyContextProvider`. I can get the unparsed result like this:
```
unparsed to SQL:
SELECT make_array(1, 2, 3)
```

However, I think `make_array` isn't a common function for standard SQL. It may cause some problems when we try to fully push down the unparsed result to a specific data source.

### Describe the solution you'd like

Ideally, we can round-trip the array construction in SQL `select [1,2,3]` through the following steps: 
1. Transform the `AST::Expr::Array` to `ScalarValue::List` (currently, it's the `make_array` scalar function).
2. Implement the unparsing of `ScalarValue::List`.

However, I'm not sure why the array expression was disabled and `make_array` is used instead. Maybe it is due to some performance issues?

### Describe alternatives you've considered

Another approach is to try unparsing the `make_array()` plan to `AST::Expr::Array`. It's tricky, but I think it could work.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #11144.

## Rationale for this change
DataFusion will plan the array construction and the elements' access to the corresponding scalar functions. If we don't handle those cases, they will be unparsed to `ast::Expr::ScalarFunction` which isn't a common SQL syntax.
### Array Construction
```
SELECT [1, 2, 3] -> make_array(1, 2, 3)
```
### Elements accessing
```
SELECT [1, 2, 3][1] -> array_element(make_array(1,2,3), 1)
```

This PR tries to allow the array SQL to do SQL-Plan-SQL roundtrip.

## What changes are included in this PR?
- Unparse `make_array` to `ast::Expr::Array`
- Unparse `array_element` to `ast::Expr::Subscript`

## Are these changes tested?
expr unit tests and SQL roundtrip tests.

## Are there any user-facing changes?
no

### CODE REVIEW

File: datafusion/sql/src/unparser/expr.rs [hunk 1]
```diff
@@ -476,6 +476,19 @@ impl Unparser<'_> {
         &self,
         func_name: &str,
         args: &[Expr],
+    ) -> Result<ast::Expr> {
+        match func_name {
+            "make_array" => self.make_array_to_sql(args),
+            "array_element" => self.array_element_to_sql(args),
+            // TODO: support for the construct and access functions of the `map` and `struct` types
```
Comment (goldmedal) [inline datafusion/sql/src/unparser/expr.rs:483]: Similar cases exist for `map` and `struct`. I'll file another issue for them.

File: datafusion/sql/tests/cases/plan_to_sql.rs [hunk 2]
```diff
@@ -182,6 +184,11 @@ fn roundtrip_statement() -> Result<()> {
             SUM(id) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS running_total
             FROM person
             GROUP BY GROUPING SETS ((id, first_name, last_name), (first_name, last_name), (last_name))"#,
+            "SELECT ARRAY[1, 2, 3]",
```
Comment (alamb) [inline datafusion/sql/tests/cases/plan_to_sql.rs:187]: 💯

**Review by alamb — APPROVED** [review]
Makes sense to me -- thank you @goldmedal

**Comment by alamb** [comment]
Thanks again @goldmedal

**Comment by goldmedal** [comment]
Thanks @alamb

**Pull request merged**